### PR TITLE
add inner_child_properties tool

### PR DESCRIPTION
**Pull request by pau557**
This tool will descend and find the properties of the inner-most children sampler in a composite (or just a sampler).

Useful for extracting QPU parameters when the QPU sampler is wrapped on many composite layers.

### CODE REVIEW

File: dimod/utilities.py [hunk 1]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
...
+
+    Returns:
+        properties (dict): The properties of the inner-most sampler
+
+    Example:
+    >>> sampler = EmbeddingComposite(ScaleComposite(qpu_sampler))
+    >>> inner_child_properties(sampler)["annealing_time_range]
+    [0.5, 2000.0]
+    """
+
```
Comment (arcondello) [inline dimod/utilities.py:468]: maybe
```python
try:
    return inner_child_properties(sampler.child)
except AttributeError:
    return sampler.properties
```

File: dimod/utilities.py [hunk 2]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
 
     raise ValueError("no structured sampler found")
 
+def inner_child_properties(sampler):
+    """
+    Returns the properties of inner-most child sampler in a composite.
```
Comment (arcondello) [inline dimod/utilities.py:455]: This should be on the first line

File: dimod/utilities.py [hunk 3]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
 
     raise ValueError("no structured sampler found")
 
+def inner_child_properties(sampler):
```
Comment (arcondello) [inline dimod/utilities.py:453]:
```python
def inner_child_properties(sampler: dimod.Sampler) -> typing.Dict[str, typing.Any]:
```

**Review by arcondello — COMMENTED** [review]
Also, maybe `innermost_child_properties` would be more explicit?

Or, another pattern,
```python
sampler = ScaleComposite(ExactSolver())

sampler.innermost_child().properties

```

File: dimod/utilities.py [hunk 4]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
...
+    Returns the properties of inner-most child sampler in a composite.
+
+    Args:
+        sampler: A dimod sampler
+
+    Returns:
+        properties (dict): The properties of the inner-most sampler
+
+    Example:
+    >>> sampler = EmbeddingComposite(ScaleComposite(qpu_sampler))
```
Comment (JoelPasvolsky) [inline dimod/utilities.py:464]: Having `EmbeddingComposite` means a dependency on dwave-system, so you'd need to skip the doctest

File: dimod/utilities.py [hunk 5]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
...
+    >>> inner_child_properties(sampler)["annealing_time_range]
```
Comment (JoelPasvolsky) [inline dimod/utilities.py:465]: Should be inner_child_properties(sampler)["annealing_time_range"]

File: dimod/utilities.py [hunk 6]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
...
+    [0.5, 2000.0]
+    """
+
+    def get_properties(properties):
+        if "child_properties" in properties:
+            return get_properties(properties["child_properties"])
+        else:
+            return properties
+    return get_properties(sampler.properties)
```
Comment (JoelPasvolsky) [inline dimod/utilities.py:474]: Some of our composites are tricky:
```
>>> sampler = EmbeddingComposite(TilingComposite(DWaveSampler(), 1, 1, 4))
>>> inner_child_properties(sampler["annealing_time_range"])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-e57179dc0149> in <module>
----> 1 inner_child_properties(sampler["annealing_time_range"])

TypeError: 'EmbeddingComposite' object is not subscriptable
```

File: dimod/utilities.py [hunk 7]
```diff
@@ -449,6 +450,28 @@ def child_structure_dfs(sampler, seen=None):
...
+    >>> sampler = EmbeddingComposite(ScaleComposite(qpu_sampler))
+    >>> inner_child_properties(sampler)["annealing_time_range]
+    [0.5, 2000.0]
+    """
+
+    def get_properties(properties):
+        if "child_properties" in properties:
+            return get_properties(properties["child_properties"])
+        else:
+            return properties
```
Comment (JoelPasvolsky) [inline dimod/utilities.py:473]: I'd add a space line here

File: dimod/utilities.py (hunk 2 above)
Comment (JoelPasvolsky) [inline dimod/utilities.py:455]: "of the inner-most child"

File: dimod/utilities.py (hunk 6 above)
Comment (pau557) [reply dimod/utilities.py:474]: Try:

```inner_child_properties(sampler)["annealing_time_range"]```

Works for me.

Comment (JoelPasvolsky) [reply dimod/utilities.py:474]: Yup, user error, sorry

File: dimod/utilities.py (hunk 4 above)
Comment (pau557) [reply dimod/utilities.py:464]: is that a problem if it's in the example docstring? It just seems like the most obvious usage to me.

I wasn't aware that docstrings get tested, does that mean that they should include imports and be well self-contained?

File: dimod/utilities.py (hunk 3 above)
Comment (pau557) [reply dimod/utilities.py:453]: ugh

**Comment by pau557** [comment]
Your suggestion makes sense @arcondello, worth just closing this PR and making a new one instead of making a complete overhaul

**Pull request closed without merging**